pkg/owner/owner.go

```diff
@@ -0,0 +1,97 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package owner contains code for walking up the ownership chain,
+// starting with an arbitrary object. RBAC rules must allow GET access
+// to each object on the chain, at least including the starting
+// object, more when walking up more than one level.
+package owner
+
+import (
+	"context"
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Lookup walks up the ownership chain zero or more levels and returns an OwnerReference for the
+// object. The object identified by name, namespace and type is the starting point and is
+// returned when levels is zero. Only APIVersion, Kind, Name, and UID will be set.
+// IsController is always true.
+func Lookup(config *rest.Config, namespace, name string, gkv schema.GroupVersionKind, levels int) (*metav1.OwnerReference, error) {
```

Comment:
Please add an unit tests with a Deployment + StatefulSet

Reply:
Even if that adds yet another dependency, specifically on sigs.k8s.io/controller-runtime/pkg/client/fake?

 I wasn't quite sure whether that would be worthwhile, but I can add it.

Reply:
It won't cover RBAC rules, though.

Reply:
Done and pushed.

Reply:
Awesome!